### PR TITLE
Issue #107 #119: HomeObject to report Per PG stats and Per-HomeObject Stats

### PR DESCRIPTION
**Pull request by yamingk**
Add per PG stats according to: https://github.corp.ebay.com/SDS/nuobject_proto/pull/10

This PR will need homestore PR: https://github.com/eBay/HomeStore/pull/243/files

A new test case added for PGstats and HomeObj stats

Test:
====
```
[12/05/23 11:26:27.953] [./bin/homestore_test] [info] [82978] [hs_blob_tests.cpp:329:TestBody] stats: PGStats: id=1, replica_set_uuid=3524160a-c694-4120-9a1c-4a7626e561c2, 
num_members=3, total_shards=2, open_shards=1, avail_open_shards=127, used_bytes=26402816, avail_bytes=1676314624, 
members: 
member-1: id=9efb16ac-964d-4bc7-883c-9f8bb363839c, name=peer1, 
member-2: id=82008433-de79-4c89-bdb5-71356cb849ae, name=peer3, 
member-3: id=a75534a5-6d20-42e3-9bdc-eaa826682d2d, name=peer2
[12/04/23 12:28:17.833] [./bin/homestore_test] [info] [80734] [hs_blob_tests.cpp:338:TestBody] HomeObj stats: total_capacity_bytes=1689518080, used_capacity_bytes=5120, num_open_shards=1, avail_open_shards=127
```

### CODE REVIEW

File: src/lib/homestore_backend/hs_pg_manager.cpp [hunk 1]
```diff
@@ -193,4 +206,73 @@ void HSHomeObject::persist_pg_sb() {
...
+    stats.id = hs_pg->pg_info_.id;
+    stats.replica_set_uuid = hs_pg->pg_info_.replica_set_uuid;
+    stats.num_members = hs_pg->pg_info_.members.size();
+    stats.total_shards = hs_pg->total_shards();
+    stats.open_shards = hs_pg->open_shards();
+
+    for (auto const& m : hs_pg->pg_info_.members) {
+        stats.members.emplace_back(m.id, m.name);
+    }
+
```
Comment (xiaoxichen) [inline src/lib/homestore_backend/hs_pg_manager.cpp:226]: I think we can add a `hs_pg->avail_chunks()` which calls `chunk_selector().avail_chunks(pdev_id)`
and
```
uint32_t HeapChunkSelector:: avail_chunks(pdev_id) {
            return  m_per_dev_heap[pdev_id].second().size();
}
```

File: src/lib/homestore_backend/hs_pg_manager.cpp [hunk 2]
```diff
@@ -193,4 +206,73 @@ void HSHomeObject::persist_pg_sb() {
...
+    auto hs_pg = static_cast< HS_PG* >(pg.get());
+    auto const blk_size = hs_pg->repl_dev_->get_blk_size();
+
+    stats.id = hs_pg->pg_info_.id;
+    stats.replica_set_uuid = hs_pg->pg_info_.replica_set_uuid;
+    stats.num_members = hs_pg->pg_info_.members.size();
+    stats.total_shards = hs_pg->total_shards();
+    stats.open_shards = hs_pg->open_shards();
+
+    for (auto const& m : hs_pg->pg_info_.members) {
```
Comment (xiaoxichen) [inline src/lib/homestore_backend/hs_pg_manager.cpp:223]: the quorum in PGStat need `up to date PG members`.  The initiative of reporting this to CM is let client (GW) "best effort" avoid read send to a "significant laggy" member,  especially , those members are under snapshot based resync.

We can only get the member stats from leader so we may only report this in leader PG. 
Also the `Quorum` struct need leader information which is not in pg_stat :)

Comment (yamingk) [reply src/lib/homestore_backend/hs_pg_manager.cpp:223]: good catch, it should be inside HomeObj stats.

**Review by xiaoxichen — APPROVED** [review]
LGTM.

File: src/lib/homestore_backend/hs_pg_manager.cpp (hunk 2 above)
Comment (xiaoxichen) [reply src/lib/homestore_backend/hs_pg_manager.cpp:223]: as raft part is not yet merged, maybe we merge this PR first and create an issue for the PG member change?

**Review by zichanglai — APPROVED** [review]
LGTM

File: src/lib/homestore_backend/hs_pg_manager.cpp (hunk 2 above)
Comment (yamingk) [reply src/lib/homestore_backend/hs_pg_manager.cpp:223]: yes, I added the required fields into the stats and left that field as zero for now. I will create another ticket to track this issue once repl_dev is ready to get the api for last committed lsn.

**Pull request merged**